web-server-lib/web-server/private/timer.rkt

```diff
@@ -18,73 +18,84 @@
 ;; The application may further adjust its allotted timeout for handling
 ;; requests by using the dispatcher in the dispatch-timeout module.
 
-(struct timer-manager (thread timer-ch))
-(define-struct timer (tm evt expire-seconds action)
-  #:mutable)
+(define-logger web-server/timer)
+
+(struct timer-manager (thd))
+(struct timer (tm [deadline #:mutable] [action #:mutable])
+  #:extra-constructor-name make-timer)
 
 ;; start-timer-manager : -> timer-manager?
 ;; The timer manager thread
 (define (start-timer-manager)
-  (define timer-ch (make-async-channel))
-  (timer-manager
-   (thread
-    (lambda ()
-      (let loop ([timers null])
-        ;; (printf "Timers: ~a\n" (length timers))
-        ;; Wait for either...
-        (apply sync
-               ;; ... a timer-request message ...
-               (handle-evt
-                timer-ch
-                (lambda (req)
-                  ;; represent a req as a (timer-list -> timer-list) function:
-                  ;; add/remove/change timer evet:
-                  (loop (req timers))))
-               ;; ... or a timer
-               (map (lambda (timer)
-                      (handle-evt
-                       (timer-evt timer)
-                       (lambda (_)
-                         ;; execute timer
-                         ((timer-action timer))
-                         (loop (remq timer timers)))))
-                    timers)))))
-   timer-ch))
-
-;; Limitation on this add-timer: thunk cannot make timer
-;;  requests directly, because it's executed directly by
-;;  the timer-manager thread
+  (define thd
+    (thread
+     (lambda ()
+       (define timers (make-hash))
+       (let loop ([first-deadline +inf.0]
+                  [deadline-evt never-evt])
+         (log-web-server/timer-debug "timers: ~a, next deadline in ~.sms" (hash-count timers) (- first-deadline (current-inexact-milliseconds)))
+         (sync
+          (handle-evt
+           (thread-receive-evt)
```

Comment:
It would be nice if this produced the message, instead of the event, so that you didn't have to deal with the thread system twice in this code.

Reply:
Yeah, I've written enough code like this now that I'm used to it, but the behavior of `thread-receive-evt` was definitely surprising initially. Looking at `thread/thread.rkt`, it wouldn't be hard to add a variant that dequeues a message, but I haven't been able to come up with a good name for it. `thread-dequeue-evt` seems like it would cause confusion.